### PR TITLE
Support Pydantic 

### PR DESCRIPTION
**Pull request by jxnl**

### CODE REVIEW

File: inference.py [hunk 1]
```diff
@@ -1,128 +1,97 @@
 import torch
-from transformers import AutoTokenizer, AutoModelForCausalLM
+from typing import List, Optional
+from transformers import LlamaTokenizer, LlamaForCausalLM
 
+from openai_types import FunctionCall, Function, TurnMessage
 from schema import generate_schema_from_functions
 
-SYSTEM_MESSAGE = """A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate input when necessary"""
+SYSTEM_MESSAGE = """A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate and correct input when necessary"""
```
Comment (jxnl) [inline inference.py]:
```suggestion
SYSTEM_MESSAGE = """A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. The assistant calls functions with appropriate input when necessary"""
```

File: inference.py [hunk 2]
```diff
@@ -1,128 +1,97 @@
...
-
-    elif message["role"] == "assistant":
-        text = "assistant:\n{content}\n".format(content=message.get("content", ""))
-
-    else:
-        raise ValueError(f'Unsupported role: {message["role"]}')
-
-    input_ids = tokenizer(text, add_special_tokens=False, return_tensors="pt").input_ids.to("cuda:0")
-    return input_ids
+def to_tokens(message: TurnMessage, tokenizer: LlamaTokenizer):
```
Comment (musabgultekin) [inline inference.py]: "tokenize" might be a better name

File: inference.py [hunk 3]
```diff
@@ -1,128 +1,97 @@
...
+                )
+        elif message.role == "function":
+            all_messages.append(
+                TurnMessage(
+                    role="function",
+                    name=f"functions.{message.name}",
+                    content=message.content,
+                )
+            )
+        all_messages.append(message)
```
Comment (musabgultekin) [inline inference.py]: This should be in else block, as this will duplicate function and assistant messages

File: inference.py [hunk 4]
```diff
@@ -1,128 +1,97 @@
...
+    ]
     return torch.cat(all_input_ids, dim=-1)
 
 
-def generate(model, tokenizer, messages, functions=None, temperature=0.7, max_new_tokens=256):
-    inputs = prepare_messages_for_inference(tokenizer, messages, functions)
-    generate_ids = model.generate(inputs, max_new_tokens=max_new_tokens, temperature=temperature)
-    generated_content = tokenizer.batch_decode(generate_ids[:, inputs.shape[1]:], skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
-    #print(generated_content)
+def generate_models(
```
Comment (musabgultekin) [inline inference.py]: maybe a better name like generate_message ? Or something else. I believe "generate" is good anyways

File: openai_types.py
Comment (musabgultekin) [inline openai_types.py:1]: Can we remove "openai" from the filename, maybe "types.py" or something else

File: server.py
Comment (musabgultekin) [inline server.py:1]: Isnt it better to put server related pydantic models to this file? rather than all types into seperate file?

**Review by musabgultekin — CHANGES_REQUESTED** [review]

**Pull request merged**